components/brave_wallet/common/zcash_utils.cc

```diff
@@ -93,4 +168,59 @@ std::vector<uint8_t> ZCashAddressToScriptPubkey(const std::string& address,
   return data;
 }
 
+// https://zips.z.cash/zip-0316#encoding-of-unified-addresses
+std::optional<std::string> ExtractTransparentPart(
+    const std::string& unified_address,
+    bool is_testnet) {
+  auto bech_result = bech32::Decode(unified_address);
+
+  if (bech_result.encoding != bech32::Encoding::BECH32M) {
+    return std::nullopt;
+  }
+
+  std::string expected_hrp = is_testnet ? "utest" : "u";
+  if (bech_result.hrp != expected_hrp) {
+    return std::nullopt;
+  }
+
+  std::vector<uint8_t> data;
+  if (!ConvertBits<5, 8, false>([&](unsigned char c) { data.push_back(c); },
+                                bech_result.data.begin(),
+                                bech_result.data.end())) {
+    return std::nullopt;
+  }
+
+  auto reverted = RevertF4Jumble(data);
+  // HRP with 16 bytes padding which is appended to the end of message
+  if (!reverted || reverted->size() < kPaddedHrpSize) {
+    return std::nullopt;
+  }
+
+  std::vector<uint8_t> padded_hrp(kPaddedHrpSize, 0);
+  std::copy(reinterpret_cast<const uint8_t*>(expected_hrp.c_str()),
+            reinterpret_cast<const uint8_t*>(expected_hrp.c_str()) +
+                expected_hrp.length(),
+            padded_hrp.begin());
+
+  // Check that HRP is similar to the padded HRP
+  if (!std::equal(padded_hrp.begin(), padded_hrp.end(),
+                  reverted->end() - kPaddedHrpSize)) {
+    return std::nullopt;
+  }
+
+  auto parts = ParseUnifiedAddress(
+      base::make_span(*reverted).subspan(0, reverted->size() - kPaddedHrpSize));
+  if (!parts.has_value()) {
+    return std::nullopt;
```

Comment:
why are we returning base::unexpected from ParseUnifiedAddress and then throwing away the error information and just returning nullopt here?

Reply:
I think we can switch ParseUnifiedAddress to return optional type since error is not propagated

Reply:
Done